### PR TITLE
Display image metadata

### PR DESCRIPTION
**Pull request by mneuwert**
## Description
Extension built allowing to view metadata (EXIF, IPTC) on image files. Licensing: IAP.

## Related Issue
#679 

## Motivation and Context
Useful feature for users who are serious about photography.

## How Has This Been Tested?
Tested on different JPEG, TIFF, PNG, RAW image files.

## Screenshots (if appropriate):
Simulator Screen Shot - iPhone Xs - 2020-07-21 at 18 44 12

Simulator Screen Shot - iPhone Xs - 2020-07-21 at 18 37 26

Simulator Screen Shot - iPhone Xs - 2020-07-21 at 18 38 01 2

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

- [X] (1) Card does not include feature menu https://github.com/owncloud/ios-app/pull/750#issuecomment-679051019 [FIXED]
- [X] (2) Many decimals in Altitude https://github.com/owncloud/ios-app/pull/750#issuecomment-679054253 [DONE]
- [X] (3) Exposure bias duplicated https://github.com/owncloud/ios-app/pull/750#issuecomment-679055999 [FIXED]

### CODE REVIEW

File: ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift [hunk 1]
```diff
@@ -46,19 +48,22 @@ public extension OCLicenseManager {
...
 		register(OCLicenseProduct(identifier: .singleDocumentScanner, name: documentScannerFeature.localizedName!, description: documentScannerFeature.localizedDescription, contents: [.documentScanner]))
 		register(OCLicenseProduct(identifier: .singleShortcuts, name: shortcutsFeature.localizedName!, description: shortcutsFeature.localizedDescription, contents: [.shortcuts]))
 		register(OCLicenseProduct(identifier: .singleDocumentMarkup, name: documentMarkupFeature.localizedName!, description: documentMarkupFeature.localizedDescription, contents: [.documentMarkup]))
+		register(OCLicenseProduct(identifier: .singlePhotoProFeatures, name: photoProFeature.localizedName!, description: photoProFeature.localizedDescription, contents: [.photoProFeatures]))
 
 		// - Subscription
-		register(OCLicenseProduct(identifier: .bundlePro, name: "Pro Features".localized, description: "Unlock all Pro Features.".localized, contents: [.documentScanner, .shortcuts, .documentMarkup]))
+		register(OCLicenseProduct(identifier: .bundlePro, name: "Pro Features".localized, description: "Unlock all Pro Features.".localized, contents: [.documentScanner, .shortcuts, .documentMarkup, .photoProFeatures]))
```
Comment (felix-schwarz) [inline ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift:66]: Just quickly glossed over the code setting up licensing for the photo pro features and noticed that one piece of the puzzle is still missing: if Photo Pro Features should be available as a one-time purchase, too, its app store non-consumable IAP would need to be provided to `OCLicenseAppStoreProvider` a few lines down from here, in `OCLicenseManager+Setup.swift:71`++.

Comment (mneuwert) [reply ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift:66]: @felix-schwarz good catch.. indeed it was missing.. fixed

File: ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import ownCloudSDK
+
+class DisplayExifMetadataAction : Action {
+	override class var identifier : OCExtensionIdentifier? { return OCExtensionIdentifier("com.owncloud.action.show-exif") }
+	override class var category : ActionCategory? { return .normal }
+	override class var name : String? { return "Image Metadata".localized }
+	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreItem, .moreFolder, .contextMenuItem] }
+	class var supportedMimeTypes : [String] { return ["image"] }
+	override class var licenseRequirements: LicenseRequirements? { return LicenseRequirements(feature: .photoProFeatures) }
```
Comment (hosy) [inline ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift:27]: Please add a keyboard shortcut for this action

File: ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+	// MARK: - Extension matching
+	override class func applicablePosition(forContext: ActionContext) -> ActionPosition {
+		if forContext.items.count == 1, let item = forContext.items.first, item.type == .file, let mimeType = item.mimeType {
+			if supportedMimeTypes.filter({ mimeType.contains($0) }).count > 0 {
```
Comment (hosy) [inline ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift:32]: Is this action available for all image types (gif, png, svg) or only for (jpeg, heic, raw) types?

File: ownCloud/Client/Actions/ImageMetadataViewController.swift [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+
+import UIKit
+import CoreLocation
+import MapKit
+import Contacts
+import ownCloudSDK
+
+extension CLLocation {
+	var dmsLatitude: String {
+		let direction = self.coordinate.latitude >= 0 ? "N" : "S"
```
Comment (hosy) [inline ownCloud/Client/Actions/ImageMetadataViewController.swift:27]: needs to be localized

File: ownCloud/Client/Actions/ImageMetadataViewController.swift [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+		return dms(from: self.coordinate.latitude) + "\"\(direction)"
+	}
+
+	var dmsLongitude: String {
+		let direction = self.coordinate.longitude >= 0 ? "E" : "W"
```
Comment (hosy) [inline ownCloud/Client/Actions/ImageMetadataViewController.swift:32]: needs to be localized

File: ownCloud/Client/Actions/ImageMetadataViewController.swift [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+	private lazy var displayDateFormatter: DateFormatter = {
+		let formatter = DateFormatter()
+		formatter.dateStyle = .medium
+		formatter.timeStyle = .medium
+		formatter.locale = Locale.current
+		return formatter
+	}()
+
+	enum MetadataParseError : Error {
+		case InvailidInput
```
Comment (hosy) [inline ownCloud/Client/Actions/ImageMetadataViewController.swift:77]: wrong spelling of `Invalid`

File: ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift (hunk 2 above)
Comment (mneuwert) [reply ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift:32]: @hosy Done.. chosen CMD + I

File: ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift (hunk 3 above)
Comment (mneuwert) [reply ownCloud/Client/Actions/Actions+Extensions/DisplayExifMetadataAction.swift]: @hosy Excluded SVG and GIF which indeed don't make sense here

File: ownCloud/Client/Actions/ImageMetadataViewController.swift (hunk 4 above)
Comment (mneuwert) [reply ownCloud/Client/Actions/ImageMetadataViewController.swift]: @hosy done

File: ownCloud/Client/Actions/ImageMetadataViewController.swift (hunk 5 above)
Comment (mneuwert) [reply ownCloud/Client/Actions/ImageMetadataViewController.swift]: @hosy done

File: ownCloud/Client/Actions/ImageMetadataViewController.swift (hunk 6 above)
Comment (mneuwert) [reply ownCloud/Client/Actions/ImageMetadataViewController.swift]: @hosy fixed

**Comment by jesmrec** [comment]
In QA stage, i will compare the values displayed in this view with the ones any other photography app will return. I will not deep dive into the values themselves, assuring that they are metadata included in the file.

My focus will be in the way the feature is displayed and sense of the displayed data. If you have any other idea about this, please let me know to take it in account.

**Comment by jesmrec** [comment]
### (1)

The item card is showing the "Pro Features" menu instead of the "Image Metadata" menu.

I mean: in the file list of a non-enterprise and non-purchased-apps server, open the card for a image file and compare the 
"Markup" and "Image Metadata"

| Markup| Image Metadata |
|:-:|:-:|
|Screen Shot 2020-08-24 at 12 35 38|Screen Shot 2020-08-24 at 12 35 47|

iPhoneXR v13.6
commit `4c50d61`

**Comment by jesmrec** [comment]
### (2)

Tiny improvement, if posible:

Screenshot 2020-08-24 at 12 45 50

i think, no more that two decimals are needed

iPhoneXR v13.6
commit `4c50d61`

**Comment by jesmrec** [comment]
### (3)

The `Capture Settings` section contains a doubled `Exposure bias` section with two different values:

Screenshot 2020-08-24 at 12 48 53

iPhoneXR v13.6
commit `4c50d61`

**Comment by mneuwert** [comment]
### Re: (1), (2) and (3)
- (2) and (3) are fixed.
- (1) I need to check, It is actually correct that "Pro Features" package is displayed as it includes all pro features. Then there is another IAP product planned for Pro Photo stuff (EXIF, RAW upload). Probably here the product definition is missing in AppStoreConnect. Need to check with @hosy

**Comment by mneuwert** [comment]
@jesmrec Now, after updating of IAP product in AppStore Connect, (1) is also fixed.

IMG_6225

**Comment by jesmrec** [comment]
Thanks, (1), (2) and (3) are fixed

**Comment by jesmrec** [comment]
Approved. Please fix conflicts and change base branch.

**Pull request merged**